### PR TITLE
Contributing.md

### LINKED ISSUE #555: Define and enforce standard derives
Many standard traits can be easily derived but are often forgotten. In #511 the discussion of what to derive arose for `Error` types in particular. We should define a standard set of traits to implement or derive for every struct/enum to make them interoperable with algorithms and data structures from std. Ideally we could enforce that via CI too, either with some heuristic using regex or something fancy like rustql (rust-corpus/rustql#6).

### PR DESCRIPTION
**Pull request by dr-orlovsky**
We've being discussing that a `CONTRIBUTING.md` guidelines are needed for some time and I took an effort to propose it. The main text is taken from `rust-lightning` version (with some additions), but it is extended on the topics we were discussing in multiple issues previously:
- list of repository maintainers
- standard derives - #555
- naming conventions (after Bitcoin Core)
- MSRV and 2018 update #510
- code formatting with `rustfmt` #172

Each of these issue-related amendments to the basic (modified) version of rust-lightning contributing guidelines are made with a separate commit.

### CODE REVIEW

File: CONTRIBUTING.md
Comment (Kixunil) [inline CONTRIBUTING.md:198]: This should also forbid traits that conflict with invariants or guarantees of the type. Examples:

* structs with `Drop` (well, this is caught by the compiler but it's not **member field** restriction)
* Types that don't have reflexive equality/ordering (no `Eq`, `Ord`)

Possibly also forbid impls with foot-gunny or non-obvious implementations.

Also forbid **deriving** `Debug` on secrets - implement it manually by censoring the secret.

File: CONTRIBUTING.md
Comment (Kixunil) [inline CONTRIBUTING.md:254]: Is it? At least `electrs` depends on it and I consider it in-production.

File: CONTRIBUTING.md
Comment (Kixunil) [inline CONTRIBUTING.md:268]: Maybe add a requirement to test `unsafe` code with sanitizers including Miri?

File: CONTRIBUTING.md
Comment (Kixunil) [inline CONTRIBUTING.md:153]: Perhaps mention that Rust-idiomatic style is preferred to "how it looks in Core" if everyone agrees.

File: CONTRIBUTING.md
Comment (TheBlueMatt) [inline CONTRIBUTING.md:186]: Maybe also mention this is only reasonable for "small" structs (in addition to the no-heap requirement).

File: CONTRIBUTING.md
Comment (TheBlueMatt) [inline CONTRIBUTING.md:233]: 1.29 :)

File: CONTRIBUTING.md
Comment (TheBlueMatt) [inline CONTRIBUTING.md:248]: Bitcoin :) and also below.

File: CONTRIBUTING.md
Comment (TheBlueMatt) [reply CONTRIBUTING.md:254]: This appears to have been copied from LNPBP and isn't right.

File: CONTRIBUTING.md
Comment (Kixunil) [reply CONTRIBUTING.md:186]: I'm not really sure if it helps in practice. Moves are `memcpy` anyway.

File: CONTRIBUTING.md
Comment (dr-orlovsky) [reply CONTRIBUTING.md:233]: Yeah, it feels like it is 0.29 :D

File: CONTRIBUTING.md
Comment (dr-orlovsky) [reply CONTRIBUTING.md:268]: Do we allow `unsafe` at all here?

Comment (dr-orlovsky) [reply CONTRIBUTING.md:287]: Added a dedicated section addressing `unsafe`: https://github.com/rust-bitcoin/rust-bitcoin/pull/587/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055R262-R266

File: CONTRIBUTING.md
Comment (dr-orlovsky) [reply CONTRIBUTING.md:204]: Added two points below addressing this issues: https://github.com/rust-bitcoin/rust-bitcoin/pull/587/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055R206-R215

File: CONTRIBUTING.md
Comment (dr-orlovsky) [reply CONTRIBUTING.md:154]: Done below

**Comment by dr-orlovsky** [comment]
@TheBlueMatt @Kixunil addressed your comments with a separate commit 4e64db1e316ecd02d5d585f5561d90dd5544a981

File: CONTRIBUTING.md
Comment (Kixunil) [inline CONTRIBUTING.md:215]: I think "... implementation excluding private data should be ..." would be clearer

File: CONTRIBUTING.md
Comment (Kixunil) [inline CONTRIBUTING.md:210]: You mean manual *implementation*?

File: CONTRIBUTING.md
Comment (Kixunil) [inline CONTRIBUTING.md:274]: I think vulnerabilities in the library should be disclosed to the library authors as well. Contact information (with PGP key and cross-references to keybase/whatever) should be provided.

Comment (sgeisler) [reply CONTRIBUTING.md:274]: Good question, currently we don't have a process for that I think. It seems like something that warrants some discussion. Are you proposing putting a list of email addresses+GPG keys into this document?

**Review by sgeisler — COMMENTED** [review]
Left some minor nits, looks good overall. Thx for tackling this issue :)

The only major thing we should talk about more are security disclosures imo.

File: CONTRIBUTING.md
Comment (Kixunil) [reply CONTRIBUTING.md:274]: I believe the standard is to put it into `SECURITY.md`, but at least linking it here would be useful.

File: CONTRIBUTING.md
Comment (dr-orlovsky) [reply CONTRIBUTING.md:184]: Since it will be valid once we will reformat the crate with `rustfmt`, so after that this part can be just uncomented and not written from scratch

**Comment by dr-orlovsky** [comment]
@Kixunil, @sgeisler updated with security suggestions

**Comment by Kixunil** [comment]
ACK

File: Cargo.toml
Comment (sgeisler) [inline Cargo.toml:27]: This seems very unrelated in this PR/commit.

**Review by sgeisler — COMMENTED** [review]
General ACK, just please revert the changes to the dependency versions (was probably just an oversight).

File: Cargo.toml
Comment (dr-orlovsky) [reply Cargo.toml]: Removed. It was not an oversight, but just me committing into the wrong branch (version update is required for error derives, which is parallel PR)

**Review by sgeisler — APPROVED** [review]
ACK 92bb0e503f34813ddb90bc4ad26a5a81d79571c4

File: CONTRIBUTING.md
Comment (apoelstra) [inline CONTRIBUTING.md:110]: To be clear, each commit must compile and pass unit tests with no errors, with every feature combination (excepting `fuzztarget`, but including compiling the fuzztests) on some reasonably recent compiler. The tip of any PR branch must also compile and pass tests with no errors on 1.29.

I don't feel very strongly about insisting that commits include their own tests. I would mildly prefer that tests be in separate commits so that it's easy to reorder them during review and check that the new tests fail without the new code.

File: CONTRIBUTING.md
Comment (apoelstra) [inline CONTRIBUTING.md:137]: Spelling, and also I'd rather drop this whole section. To become a maintainer I need to approve you and I have no explicit policy.

**Comment by dr-orlovsky** [comment]
@apoelstra did all of that + tried to make CI scripts to match the guidelines as you described

File: contrib/test.sh
Comment (sgeisler) [inline contrib/test.sh:10]: I wonder if we should replace `PIN_VERSIONS` with a check for the toolchain version? I removed that some time ago because the toolchain variable was used for nothing else, but now that might make sense again …

File: contrib/test.sh
Comment (sgeisler) [inline contrib/test.sh:6]: Since this PR tells devs to use this script locally too I'm not sure if we should just delete the lockfile. Maybe print a warning instead if using an old rust version? The user will notice. Or at least gate it with a flag.

File: contrib/test.sh
Comment (sgeisler) [inline contrib/test.sh:75]: Let's hope no one ever links `dep_test` to `/` :stuck_out_tongue_winking_eye:

Comment (dr-orlovsky) [reply contrib/test.sh:75]: If they do, the script will fail before, on `cargo new dep_test`

File: contrib/test.sh
Comment (dr-orlovsky) [reply contrib/test.sh:10]: This will break the script once we update MSRV soon (and of course we will forget to update it here). Also script may be used for checking with custom toolchains and we cant do a reliable detection when we need to pin the versions

File: contrib/ci.sh
Comment (dr-orlovsky) [reply contrib/ci.sh:3]: You have to thank @apoelstra who was teaching in his reviews to do this constantly :)

File: contrib/test.sh
Comment (dr-orlovsky) [reply contrib/test.sh:6]: I will add NB about this to the CONTRIBUTING

File: contrib/test.sh
Comment (sgeisler) [reply contrib/test.sh:10]: > This will break the script once we update MSRV

It should, we should unpin as many deps as possible then. But yeah, when you don't want to set the toolchain that way it's still useful to have it as a separate flag.

Comment (dr-orlovsky) [reply contrib/test.sh:10]: I occasionally run tests on 1.36 w/o pins and it failed, so pinning probably will not change...

File: contrib/test.sh
Comment (dr-orlovsky) [reply contrib/test.sh:6]: Did that below

**Review by sgeisler — APPROVED** [review]

**Review by sanket1729 — COMMENTED** [review]
Would require an update to the IRC channels.

**Comment by dr-orlovsky** [comment]
@sanket1729 did that

**Review by sanket1729 — COMMENTED** [review]
One more small nit. The rest looks good.

**Comment by dr-orlovsky** [comment]
@sanket1729 the comment does not show up, are you sure you have posted it?

**Comment by dr-orlovsky** [comment]
Rebased onto the recent master. Seems like we already had very detailed discussion on this PR and it also provides improvements in CI standardization (test scripts etc), so I ask @apoelstra to merge it if possible

**Review by sanket1729 — COMMENTED** [review]
ACK modulo squashing commits.

**Review by sanket1729 — COMMENTED** [review]
ACK e0bdaf808ae861a508af6413f097ee67927ed223. Looks good, fixups should be squashed.

**Comment by dr-orlovsky** [comment]
@sanket1729 done

**Comment by apoelstra** [comment]
Almost every line of every commit has trailing spaces.

**Comment by dr-orlovsky** [comment]
@apoelstra not sure I am getting it. Are you talking about commit messages or the .md file? Since I do not see spaces at the end of commit lines. If this is about .md file, do I need to re-do all the history and remove trailing spaces, is this critical?

**Comment by sanket1729** [comment]
ACk f198131. Reviewed that the last commit is only removing white spaces. Did not check whether it removed all of them :P

File: CONTRIBUTING.md
Comment (tobin-crypto) [inline CONTRIBUTING.md:183]:
```suggestion
which will introduce no changes other than re-formatting, such that each change
```

File: CONTRIBUTING.md
Comment (tobin-crypto) [inline CONTRIBUTING.md:310]:
```suggestion
You may be interested in the guide by Jon Atack on
```

File: CONTRIBUTING.md
Comment (Kixunil) [inline CONTRIBUTING.md:237]: Randomly noticed I forgot about not implementing `Copy` on iterators.

File: CONTRIBUTING.md
Comment (tcharding) [inline CONTRIBUTING.md:131]:
```suggestion
PR authors may also find it useful to run the following script locally in order
to check that each of the commits within the PR satisfies the requirements
above, before submitting the PR to review:
```

File: CONTRIBUTING.md
Comment (tcharding) [inline CONTRIBUTING.md:166]:
```suggestion

```

File: CONTRIBUTING.md
Comment (tcharding) [inline CONTRIBUTING.md:228]:
```suggestion
   - types which have a lexicographic ordering defined as a part of a standard
     must provide a manual implementation
   - types which may be more efficiently compared with bitcoin-specific rules
     should provide a manual implementation
```

File: CONTRIBUTING.md
Comment (tcharding) [inline CONTRIBUTING.md:292]:
```suggestion
apoelstra@wpsoftware.net, encrypted with his public key, which may be found
```

File: CONTRIBUTING.md
Comment (tcharding) [inline CONTRIBUTING.md:301]:
```suggestion
Related to the security aspect, rust bitcoin developers take testing very
seriously. Due to the modular nature of the project, writing new test cases is
easy and good test coverage of the codebase is an important goal. Refactoring
the project to enable fine-grained unit testing is also an ongoing effort.
```

**Review by tcharding — COMMENTED** [review]
Thanks @dr-orlovsky, I did a fairly grammatically picky review. My grammar is not excellent by any means so if you disagree with the suggestions feel free to ignore them. I have two additional nitpicks please
- Some times the file use two newlines between headings and sometimes a single line, can we pick one of these and be uniform across the file?
- Some lists use capitalization with ending period and some lists use no capitalization and no ending period. Can we pick one of these and be uniform also please?

Thanks

File: CONTRIBUTING.md
Comment (dr-orlovsky) [reply CONTRIBUTING.md:238]: Fixed

**Comment by dr-orlovsky** [comment]
@Kixunil, @tcharding all your suggestions are added as last two commits - to siplify re-review for this large doc for those, who already reviewed and ACKed it before.

I propose to let it go the way it is, even if there are minor grammatical nits, since we can always do them as a separate PR and it is better to have non-grammar-perfect CONTRIBUTING guidelines than not to have them at all.

@sgeisler, @sanket1729 asking to re-ACK

@apoelstra, @TheBlueMatt pls review, since without it we can't finalize & merge this PR

**Review by tcharding — COMMENTED** [review]
Great, thanks.

**Comment by Kixunil** [comment]
ACK

**Review by Kixunil — APPROVED** [review]
I've re-read this again and I believe it's solid and would be better to merge this sooner rather than later. My comment about error types should be addressed in followup PR. I've also found a few nits but I don't think they are worth our time.

**Comment by tcharding** [comment]
Friendly bump, please ack so we can merge :)

In no particular order: @sgeisler, @sanket1729, @apoelstra, @TheBlueMatt, @elichai

**Comment by apoelstra** [comment]
Please rebase to
* FIx the gazillion trailing spaces on almost every line
* Bring fixes back to the commits they affect (e.g. Libera vs Freenode, Rust-LNPBP vs rust-bitcoin)

I also really don't like the rustfmt section. It is not "planned" to use rustfmt, there is a tracking issue which highlights several serious problems with rustfmt which have not been fixed in many years, and which will clearly not be fixed because the compiler tools community is too insular.

I also don't like the Derivation section. It is hard to follow and mostly summarized with "please use common sense". I think reviewers should just check on this on a case-by-case basis. Consider all the discussion happening on PRs to add `PartialEq` to errors, for example.

We should absolutely drop "Anyone can become a maintainer if they <meet some technical goals>". There are serious culture fit requirements.

There are several paragraphs, such as 
```Overall, this library must reflect Bitcoin Core approach whenever possible.
However, since many of the things in Bitcoin Core are maintained due to 
historical reasons and may represent poor design, Rust-idiomatic style is 
preferred to "how it looks in Core" if everyone agrees.
```
which appear to say nothing at all, and should be drop. There is also overuse of passive language and general flow concerns, though I won't hold up the PR because of those.

**Comment by dr-orlovsky** [comment]
@apoelstra: I tried to address your suggestions as much as possible.

First, I separated commits unrelated to CONTRIBUTION into a new #767 - and squashed the rest. Previously, this guidelines was a copy from rust-lightning guidelines with changes on top of them - and later fixups. Since the amount of changes is already large, there is no reason for the reviewers to follow that history and we may start with a final text discussed on the previous round.

The squashing also removed "Anyone can become a maintainer if they" part which was actually already removed before in one of fixup commits.

Next, on top of a single initial commit I have applied changes that you have requested:
- Removed derivation section
- Rewrote section about formatting
- Improved style, removing passive language wherever possible.

I  propose to get this thing as an initial version and improve over it in the future.

File: CONTRIBUTING.md
Comment (apoelstra) [inline CONTRIBUTING.md:65]: In c7967c87b1856ef933ea556ebf466d68c0fd9487:

Still mentions freenode

File: CONTRIBUTING.md
Comment (apoelstra) [inline CONTRIBUTING.md:131]: In 7ac1fc5f6473838988a55b9fc65d965d3079ff46:

More trailing spaces.

Comment (dr-orlovsky) [reply CONTRIBUTING.md:131]: How do you see them?
For me they are usually all get removed by `rustfmt`, but here we do not use that - and my eyes are not as laser as yours to see them clearly :D :D

**Comment by dr-orlovsky** [comment]
@apoelstra ready for re-review

File: CONTRIBUTING.md
Comment (Kixunil) [reply CONTRIBUTING.md]: `grep ' $' README.md` should do the trick. :)

Comment (dr-orlovsky) [reply CONTRIBUTING.md]: It did, but my eyes does not run `grep` as an internal extension. I tried CRISPR for that, but grep is still missed from existing CRISPR repos for human genes :D

Comment (apoelstra) [reply CONTRIBUTING.md]: I use [this tool as my git-diff extension](https://github.com/so-fancy/diff-so-fancy) which highlights trailing whitespace in red.

I am hesitant to recommend it because of the obvious security implications of replacing your diff tool, but it seems to have been working great for me (and has never shown me a diff inconsistent with the normal diff utility, in the times that I've checked).

File: CONTRIBUTING.md
Comment (apoelstra) [inline CONTRIBUTING.md:96]: In d581c9b546fb5b049aea7ba27f24edd1c44f811a:

I think `commenters` was correct.

File: CONTRIBUTING.md
Comment (apoelstra) [inline CONTRIBUTING.md:224]: In d581c9b546fb5b049aea7ba27f24edd1c44f811a:

Typo `helipng`

File: CONTRIBUTING.md
Comment (apoelstra) [inline CONTRIBUTING.md]: In 78d1a827697a757ccccf632280787f8a9be5a534:

`features` should be `feature`

**Review by apoelstra — APPROVED** [review]
ACK e1c8e13cb89d2b3fb732d317262c763a09f922f8

File: CONTRIBUTING.md
Comment (tcharding) [reply CONTRIBUTING.md]: `git diff` shows trailing whitespace. You might need `--color=always` if you pass the output to a pager. FTR I use `vimpager` and have the following shell fuction along with a shell alias for it.
```
git-log-patch () {
    local commit
    if (( $# < 1 ))
    then
        commit="HEAD~"
    else
        commit="$1"
    fi
    git log --color=always --patch --reverse "${commit}".. | vimpager
}
```

Can be used like: `git-log-patch @~~` to get diffs of the last two patches.

**Comment by dr-orlovsky** [comment]
@Kixunil pls re-ACK. I updated the file according to Andrew's review. We have removed derive section to be added later as a separate PR after a proper discussion.

**Review by Kixunil — DISMISSED** [review]
ACK e1c8e13cb89d2b3fb732d317262c763a09f922f8

**Comment by dr-orlovsky** [comment]
@apoelstra shell we wait for more ACKs here - or this can be merged already?

**Comment by apoelstra** [comment]
Let's wait for one more ACK. It's not going to conflict with anything.

File: CONTRIBUTING.md
Comment (RCasatta) [inline CONTRIBUTING.md:50]: It looks the logs at http://gnusha.org/rust-bitcoin/ didn't switch to libera because they are not working since about June 2021

Comment (dr-orlovsky) [reply CONTRIBUTING.md:50]: Yes. We have to contact Bryan Bishop to do that.

Comment (apoelstra) [reply CONTRIBUTING.md:50]: Sure, I'll ping him.

Comment (apoelstra) [reply CONTRIBUTING.md:50]: Actually I just checked and it looks liek https://gnusha.org/bitcoin-rust/ is up to date.

So actually, we should change our CONTRIBUTING file and our IRC /topic.

Comment (dr-orlovsky) [reply CONTRIBUTING.md:50]: May I do that with a fixup commit not to make re-review the whole PR for others?

Comment (Kixunil) [reply CONTRIBUTING.md:50]: I vote for fixup, reviewing this is a huge pain already.

Comment (apoelstra) [reply CONTRIBUTING.md:50]: Yep for sure.

Comment (RCasatta) [reply CONTRIBUTING.md:50]: > Actually I just checked and it looks liek https://gnusha.org/bitcoin-rust/ is up to date.

Is it correct that days are there, but they are empty (except from log-start/log-end) up to June 2021?

Comment (apoelstra) [reply CONTRIBUTING.md:50]: There is very little activity on the channel, but the logs do not appear empty to me.

Comment (RCasatta) [reply CONTRIBUTING.md:50]: Ah sorry I was watching http://gnusha.org/rust-bitcoin/ while you linked http://gnusha.org/bitcoin-rust/ (rust and bitcoin switched)

Comment (dr-orlovsky) [reply CONTRIBUTING.md]: Done

**Comment by dr-orlovsky** [comment]
@apoelstra @Kixunil @RCasatta @sanket1729 I fixed the reference to IRC logs. Pls re-ack

**Review by Kixunil — APPROVED** [review]
ACK 31c49832b70f29f9d49ab9bb94050b68cc855f1b

**Review by apoelstra — APPROVED** [review]
ACK 31c49832b70f29f9d49ab9bb94050b68cc855f1b

**Comment by dr-orlovsky** [comment]
Sorry for disturbing, @RCasatta @sanket1729, but this needs just a review of a single fixup commit in 31c49832b70f29f9d49ab9bb94050b68cc855f1b since your last reviews of this PR. We don't want to merge without your ACKs since you did a lot of commenting here before.

**Comment by RCasatta** [comment]
ACK 31c4983

**Review by sanket1729 — APPROVED** [review]
utACK 31c49832b70f29f9d49ab9bb94050b68cc855f1b. Looks good

**Comment by dr-orlovsky** [comment]
Seems like we have 4 ACKs now (@RCasatta did it in a wrong way not clicking GitHub button, but I checked with bitcoin merge tool, it still sees/uses it). @apoelstra wouldn't you mind about merging it?

**Comment by RCasatta** [comment]
I noticed one should take care using the word ACK in phrases not meant to be for acking otherwise the tool will take it for the "ACKs for top commit:" section even if it's not intended.

**Comment by apoelstra** [comment]
I think, as long as you don't mention the top commit in the same line, you should be okay?

**Comment by RCasatta** [comment]
yes, if you don't mention the top commit you are fine, pretty rare false positive, so it happened this time :)

**Pull request merged**
